### PR TITLE
feat: Add `cp` builtin to shell for Windows

### PR DESCRIPTION
**Pull request by zackradisic**
### What does this PR do?

This adds the `cp` shell builtin to Bun shell

Not all options are supported so it is only enabled on Windows right now

### How did you verify your code works?

- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I included a test for the new code, or an existing test covers it
- [x] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [x] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)

### CODE REVIEW

**Comment by ghiscoding** [comment]
I'm assuming this PR should probably reference #9716

File: src/bun.js/node/node_fs.zig [hunk 1]
```diff
@@ -242,10 +242,79 @@ pub const Async = struct {
     }
 };
 
+pub const AsyncCPTaskVtable = struct {
```
Comment (gvilums) [inline src/bun.js/node/node_fs.zig]: I'm not sure I understand the point of the vtable here. The set of implementations is closed - either it's in js, or it's in the shell. Wouldn't a simple boolean suffice? The vtable introduces a lot of complexity, and handling checks by comparing pointers to a magic value is bad practice in my opinion.
My suggestion: move the `is_shell` boolean to the `AsyncCpTask`, and just use `if` statements instead of a vtable.

**Review by gvilums — CHANGES_REQUESTED** [review]

File: src/bun.js/node/node_fs.zig [hunk 2]
```diff
@@ -243,9 +243,9 @@ pub const Async = struct {
 };
 
 pub const AsyncCpTask = struct {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_fs.zig]: Can't we make this a comptime fn and then have `AsyncShellCpTask`?

**Comment by Jarred-Sumner** [comment]
I don't think we can ship this without `cp -r`

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
This segfaults when copying directories.

That means:
1) The automated tests are not nearly comprehensive enough
2) You didn't manually test copying directories

Please test your code both manually and through automated means.

```js
PS C:\bun> bun-10174 exec 'cp -R node_modules C:\temp\node_mods'
Segmentation fault at address 0xb0
???:?:?: 0x7ff6b7deebc3 in ??? (bun-10174.exe)
???:?:?: 0x7ffd56de257c in ??? (KERNEL32.DLL)
???:?:?: 0x7ffd5840aa47 in ??? (ntdll.dll)
```

**Comment by Jarred-Sumner** [comment]
many test failures

File: src/bun.zig [hunk 3]
```diff
@@ -768,6 +768,22 @@ pub fn openDirAbsolute(path_: []const u8) !std.fs.Dir {
...
+                is_enabled = false;
+                if (getenvZ(flag_)) |val| {
+                    is_enabled = strings.eqlComptime(val, "1") or strings.eqlComptime(val, "true");
+                }
+
+                break :brk is_enabled.?;
+            };
+        }
+    }.get();
+}
```
Comment (gvilums) [inline src/bun.zig]: this has a data race in multi-threaded code, `is_enabled` should be atomic

File: src/shell/interpreter.zig [hunk 4]
```diff
@@ -10105,6 +10118,616 @@ pub const Interpreter = struct {
...
+                fn enqueueToEventLoop(this: *ShellCpTask) void {
+                    if (this.event_loop == .js) {
+                        this.event_loop.js.enqueueTaskConcurrent(this.concurrent_task.js.from(this, .manual_deinit));
+                    } else {
+                        this.event_loop.mini.enqueueTaskConcurrent(this.concurrent_task.mini.from(this, "runFromMainThreadMini"));
+                    }
+                }
+
+                pub fn runFromMainThread(this: *ShellCpTask) void {
+                    print("runFromMainThread", .{});
```
Comment (gvilums) [inline src/shell/interpreter.zig]: is this print supposed to be here?

**Review by gvilums — COMMENTED** [review]
Looks good overall. See minor feedback related to flags processing

File: src/bun.zig (hunk 3 above)
Comment (Jarred-Sumner) [reply src/bun.zig]: its harmless if it's checked multiple times

Comment (gvilums) [reply src/bun.zig]: if it's null and two threads try to initialize it at the same time you get undefined behavior because the two writes do not synchronize

Comment (zackradisic) [reply src/bun.zig]: In the case that two threads try to initialize it, would it matter since they would both compute the same value sincee env is constant? I guess it would result in a slight performance degradation

I'll make it atomic just to be safe anyway

File: src/shell/interpreter.zig (hunk 4 above)
Comment (zackradisic) [reply src/shell/interpreter.zig]: this is the bun output scoped thing, not std.debug.print

File: src/bun.zig (hunk 3 above)
Comment (gvilums) [reply src/bun.zig]: yes it would matter, two threads writing to the same memory without synchronization is undefined behavior, no matter if they're writing the same value or not

**Comment by Jarred-Sumner** [comment]
@zackradisic there are some conflicts now due to the other PR

**Comment by Jarred-Sumner** [comment]
but its very close to mergeable

**Comment by Jarred-Sumner** [comment]
@zackradisic the next step for this PR is to fix the failing cp test on Windows

File: src/bun.js/node/node_fs.zig [hunk 5]
```diff
@@ -242,111 +242,486 @@ pub const Async = struct {
...
+                    .tracker = JSC.AsyncTaskTracker{ .id = 0 },
+                    .arena = arena,
+                    .subtask_count = .{ .raw = 1 },
+                    .shelltask = shelltask,
+                },
+            );
+            task.ref.ref(mini);
+            task.args.src.toThreadSafe();
+            task.args.dest.toThreadSafe();
+            // task.tracker.didSchedule(globalObject);
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_fs.zig]:
```suggestion

```

File: src/bun.js/node/node_fs.zig [hunk 6]
```diff
@@ -3432,11 +3741,14 @@ pub const Arguments = struct {
             recursive: bool,
             errorOnExist: bool,
             force: bool,
+            deinit_paths: bool = true,
         };
 
         fn deinit(this: Cp) void {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_fs.zig]:
```suggestion
        fn deinit(this: *Cp) void {
```

File: src/bun.js/node/node_fs.zig [hunk 7]
```diff
@@ -3432,11 +3741,14 @@ pub const Arguments = struct {
...
-            this.src.deinit();
-            this.dest.deinit();
+            if (this.flags.deinit_paths) {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_fs.zig:3747]: i don't feel good about this

File: src/shell/interpreter.zig [hunk 8]
```diff
@@ -10101,6 +10114,734 @@ pub const Interpreter = struct {
...
+        pub const Cp = struct {
+            bltn: *Builtin,
+            opts: Opts = .{},
+            state: union(enum) {
+                idle,
+                exec: struct {
+                    target_path: [:0]const u8,
+                    paths_to_copy: []const [*:0]const u8,
+                    started: bool = false,
+                    tasks_count: i32 = 0,
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]: is this thread safe? does it need to be thread safe? why is it an i32 instead of a u32?

File: src/shell/interpreter.zig [hunk 9]
```diff
@@ -10101,6 +10114,734 @@ pub const Interpreter = struct {
...
+                        .waiting_write_err => return,
+                        .done => unreachable,
+                    }
+                }
+
+                this.bltn.done(0);
+            }
+
+            pub fn deinit(this: *Cp) void {
+                _ = this; // autofix
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]: Is this intentionally missing a deinit method?

File: test/js/bun/shell/commands/cp.test.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+const p = process.platform === "win32" ? (s: string) => s.replaceAll("/", "\\") : (s: string) => s;
+
+$.nothrow();
+
+describe.if(!builtinDisabled("cp"))("bunshell cp", async () => {
+  TestBuilder.command`cat ${import.meta.filename} > lmao.txt; cp -v lmao.txt lmao2.txt`
+    .stdout(p("$TEMP_DIR/lmao.txt -> $TEMP_DIR/lmao2.txt\n"))
+    .ensureTempDir()
+    .testMini()
+    .fileEquals("lmao2.txt", await $`cat ${import.meta.filename}`.text())
```
Comment (Jarred-Sumner) [inline test/js/bun/shell/commands/cp.test.ts]: If this throws, the rest of the tests in this file will not run and we will never know

Please move all of usages of top-level await in this test to be inside closures.

File: test/js/bun/shell/commands/dirname.test.ts [hunk 11]
```diff
@@ -3,6 +3,7 @@ import { describe, test, expect } from "bun:test";
 import { createTestBuilder } from "../test_builder";
 const TestBuilder = createTestBuilder(import.meta.path);
 
+$.nothrow();
```
Comment (Jarred-Sumner) [inline test/js/bun/shell/commands/dirname.test.ts]: why?

File: test/js/bun/shell/commands/exit.test.ts [hunk 12]
```diff
@@ -5,6 +5,7 @@ const TestBuilder = createTestBuilder(import.meta.path);
 import { sortedShellOutput } from "../util";
 import { join } from "path";
 
+$.nothrow();
```
Comment (Jarred-Sumner) [inline test/js/bun/shell/commands/exit.test.ts]: why? shouldn't the TestBuilder suppress the exception if it's expected to have a non-zero exit code?

File: src/shell/interpreter.zig [hunk 13]
```diff
@@ -10101,6 +10114,734 @@ pub const Interpreter = struct {
...
+                const exit_code = this.state.ebusy.main_exit_code;
+                this.state = .done;
+                this.bltn.done(exit_code);
+            }
+
+            pub fn next(this: *Cp) void {
+                while (this.state != .done) {
+                    switch (this.state) {
+                        .idle => @panic("Invalid state for \"Cp\": idle, this indicates a bug in Bun. Please file a GitHub issue"),
+                        .exec => {
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:10258]: maybe there should be a shell panic that includes additional information like what command was run?

File: src/shell/interpreter.zig [hunk 14]
```diff
@@ -10101,6 +10114,734 @@ pub const Interpreter = struct {
...
+                cwd_path: [:0]const u8,
+                verbose_output_lock: std.Thread.Mutex = .{},
+                verbose_output: ArrayList(u8) = ArrayList(u8).init(bun.default_allocator),
+
+                task: JSC.WorkPoolTask = .{ .callback = &runFromThreadPool },
+                event_loop: JSC.EventLoopHandle,
+                concurrent_task: JSC.EventLoopTask,
+                err: ?bun.shell.ShellErr = null,
+
+                const print = bun.Output.scoped(.ShellCpTask, false);
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]: we don't use `print` anywhere else in the codebase for these

can this be called `debug` or `log`?

File: src/shell/interpreter.zig [hunk 15]
```diff
@@ -10101,6 +10114,734 @@ pub const Interpreter = struct {
...
+                    return ResolvePath.Platform.auto.isSeparator(path[path.len - 1]);
+                }
+
+                const Kind = enum {
+                    file,
+                    dir,
+                };
+
+                pub fn isDir(this: *ShellCpTask, path: [:0]const u8) Maybe(bool) {
+                    _ = this;
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]:
```suggestion

```

File: src/shell/interpreter.zig [hunk 16]
```diff
@@ -10101,6 +10114,734 @@ pub const Interpreter = struct {
...
+                    };
+                    this.onCopyImpl(src, dest);
+                }
+
+                pub fn cpOnFinish(this: *ShellCpTask, result: Maybe(void)) void {
+                    this.onSubtaskFinish(result);
+                }
+            };
+
+            const Opts = struct {
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]:
```suggestion
            const Opts = packed struct {
```

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
A few comments

File: src/shell/interpreter.zig (hunk 8 above)
Comment (zackradisic) [reply src/shell/interpreter.zig]: its thread safe as it is only read and mutated by the main thread, changed it to a u32

**Comment by Jarred-Sumner** [comment]
@zackradisic 
> [test/js/bun/shell/commands/cp.test.ts](https://github.com/oven-sh/bun/pull/10174#testjsbunshellcommandscptestts) 1 failing

File: test/js/bun/shell/test_builder.ts [hunk 17]
```diff
@@ -17,32 +18,42 @@ export function createTestBuilder(path: string) {
   });
 
   class TestBuilder {
-    promise: { type: "ok"; val: ShellPromise } | { type: "err"; val: Error };
+    // promise: { type: "ok"; val: ShellPromise } | { type: "err"; val: Error };
```
Comment (gvilums) [inline test/js/bun/shell/test_builder.ts]: can we delete this?

File: test/js/bun/shell/test_builder.ts [hunk 18]
```diff
@@ -260,63 +304,104 @@ export function createTestBuilder(path: string) {
...
+    result += characters.charAt(Math.floor(Math.random() * charactersLength));
+  }
+
+  return result;
+}
+
+export function bunExe() {
+  if (process.platform === "win32") return process.execPath.replaceAll("\\", "/");
+  return process.execPath;
+}
```
Comment (gvilums) [inline test/js/bun/shell/test_builder.ts:406]: don't we already have a `bunExe` function in `harness.ts`?

File: test/js/bun/shell/bunshell-default.test.ts [hunk 19]
```diff
@@ -1,25 +1,54 @@
...
-    await $`echo hi; ls oogabooga`.quiet();
-    expect.unreachable();
-  } catch (e: any) {
-    expect(e).toBeInstanceOf(Error);
-    expect(e.exitCode).toBe(1);
-    expect(e.message).toBe("Failed with exit code 1");
-    expect(e.stdout.toString("utf-8")).toBe("hi\n");
-    expect(e.stderr.toString("utf-8")).toBe("ls: oogabooga: No such file or directory\n");
-  }
+  // Run in a subproc because other tests may change the value of $.throws
```
Comment (gvilums) [inline test/js/bun/shell/bunshell-default.test.ts:7]: A thought: Maybe it would make sense to allow users to create a shell instance that is configurable and not global, e.g. `const $ = Bun.$.withOptions(...);`. This would alleviate the issues here. (obviously this is out of scope for this PR)

File: src/shell/subproc.zig [hunk 20]
```diff
@@ -505,7 +505,7 @@ pub const ShellSubprocess = struct {
         // }
 
         // if (!this.hasCalledGetter(.stderr)) {
-        this.stdout.unref();
+        this.stderr.unref();
         // }
```
Comment (gvilums) [inline src/shell/subproc.zig]: can we get rid of the commented-out stuff?

File: src/shell/interpreter.zig [hunk 21]
```diff
@@ -10107,6 +10120,744 @@ pub const Interpreter = struct {
...
+            /// executed, to avoid complicated synchronization on multiple
+            /// threads, because the precise src or dest for each argument is
+            /// not known until its corresponding ShellCpTask is executed by the
+            /// threadpool.
+            const EbusyState = struct {
+                tasks: std.ArrayListUnmanaged(*ShellCpTask) = .{},
+                absolute_targets: std.StringArrayHashMapUnmanaged(void) = .{},
+                absolute_srcs: std.StringArrayHashMapUnmanaged(void) = .{},
+
+                pub fn deinit(this: *EbusyState) void {
```
Comment (gvilums) [inline src/shell/interpreter.zig:10178]: do we need to free the tasks themselves?

File: src/shell/interpreter.zig [hunk 22]
```diff
@@ -10107,6 +10120,744 @@ pub const Interpreter = struct {
...
+
+            pub fn ignoreEbusyErrorIfPossible(this: *Cp) void {
+                if (!bun.Environment.isWindows) @compileError("dont call this plz");
+
+                if (this.state.ebusy.idx < this.state.ebusy.state.tasks.items.len) {
+                    outer_loop: for (this.state.ebusy.state.tasks.items[this.state.ebusy.idx..], 0..) |task_, i| {
+                        const task: *ShellCpTask = task_;
+                        assert(task.tgt_absolute != null);
+                        const failure_src = task.src_absolute.?;
+                        const failure_tgt = task.tgt_absolute.?;
```
Comment (gvilums) [inline src/shell/interpreter.zig]: why do we assert the target but not the source? do we need the assert at all if it's only going to be in debug builds anyway, which check `.?` ?

File: src/shell/interpreter.zig [hunk 23]
```diff
@@ -10107,6 +10120,744 @@ pub const Interpreter = struct {
...
+                        task.src_absolute = null;
+                        if (src_absolute) |tgt| this.state.exec.ebusy.absolute_srcs.put(bun.default_allocator, tgt, {}) catch bun.outOfMemory();
+                    }
+                }
+
+                this.printShellCpTask(task);
+            }
+
+            pub fn printShellCpTask(this: *Cp, task: *ShellCpTask) void {
+                defer task.deinit();
```
Comment (gvilums) [inline src/shell/interpreter.zig:10369]: is this `defer` correct? why are we deinitializing the task if we're starting it in all codepaths?

**Review by gvilums — APPROVED** [review]

File: test/js/bun/shell/test_builder.ts (hunk 18 above)
Comment (zackradisic) [reply test/js/bun/shell/test_builder.ts:406]: the shell leak tests do some hacky things like taking the `test_builder.ts` file and copying it, its easier to just rewrite the function here

File: test/js/bun/shell/bunshell-default.test.ts (hunk 19 above)
Comment (zackradisic) [reply test/js/bun/shell/bunshell-default.test.ts:7]: Yeah we should do this

File: src/shell/interpreter.zig (hunk 21 above)
Comment (zackradisic) [reply src/shell/interpreter.zig:10178]: added a comment they are freed in ignoreEbusyErrorIfPossible

File: src/bun.js/node/node_fs.zig [hunk 24]
```diff
@@ -242,111 +242,491 @@ pub const Async = struct {
...
+                }
+
+                this.deinit();
+            }
+
+            pub fn deinit(this: *ThisSingleTask) void {
+                // There is only one path buffer for both paths. 2 extra bytes are the nulls at the end of each
+                bun.default_allocator.free(this.src.ptr[0 .. this.src.len + this.dest.len + 2]);
+
+                bun.destroy(this);
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_fs.zig:341]: not sure why this is using `bun.destroy` / `bun.new` instead of `usingnamespace bun.New` and ThisSingleTask.new / this.destroy

File: src/bun.js/node/node_fs.zig [hunk 25]
```diff
@@ -242,111 +242,491 @@ pub const Async = struct {
...
+            this: *ThisAsyncCpTask,
+        ) void {
+            const args = this.args;
+            var src_buf: bun.OSPathBuffer = undefined;
+            var dest_buf: bun.OSPathBuffer = undefined;
+            const src = args.src.osPath(@ptrCast(&src_buf));
+            const dest = args.dest.osPath(@ptrCast(&dest_buf));
+
+            if (Environment.isWindows) {
+                const attributes = windows.GetFileAttributesW(src);
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_fs.zig:518]: I'm surprised we don't have a helper for this
